### PR TITLE
extend binomial to real and complex numbers

### PR DESCRIPTION
**Pull request by CarloLucibello**

### CODE REVIEW

File: base/special/gamma.jl [hunk 1]
```diff
@@ -482,3 +482,7 @@ eta(x::Integer) = eta(Float64(x))
 eta(x::Real)    = oftype(float(x),eta(Float64(x)))
 eta(z::Complex) = oftype(float(z),eta(Complex128(z)))
 @vectorize_1arg Number eta
+
+Base.binomial{S<:Number,T<:Number}(n::S, k::T) = ( (isinteger(n) && isinteger(k))
```
Comment (tkelman) [inline base/special/gamma.jl]: this would be better off using dispatch rather than runtime type checks

Comment (CarloLucibello) [inline base/special/gamma.jl]: here with `isinteger` I don't check on type but only if I can cast to some integer type without loss. This is to be able to compute `binomial(2.0, 1)`

Comment (tkelman) [inline base/special/gamma.jl]: ah, right. the generic `isinteger` for floating point numbers is comparing whether the value is equal to `trunc` of itself and is finite, which doesn't mean it'll necessarily fit in an `Int64` - for type stability I wonder whether it would be any simpler to just always use the definition in terms of `gamma`

Comment (CarloLucibello) [inline base/special/gamma.jl]: it would be easier but we also have to address the case where we have infinities both in the numerator and in the denominator, that result in a finite number for the fraction. This can happen only for negative integer arguments of the gamma functions, and it is handed nicely by the integer version of `binomial`

Comment (tkelman) [inline base/special/gamma.jl]: Would be ideal to have more test cases for all these corner cases. What should be done for values too large for `Int64` ?

Comment (CarloLucibello) [inline base/special/gamma.jl]: What about converting always to `BigInt` instead of `Int64`? 

```? convert(promote_type(T,S), binomial(round(BigInt, n), round(BigInt, k)))```

it would be a little slower but will address the corner case of those crazy enough to evaluate binomials with such big numbers

Comment (nalimilan) [inline base/special/gamma.jl]: The general practice in Julia is to use native `Int` by default, to avoid imposing an overhead on all use cases just to support very specific cases. Always using `BigInt` goes against this idea.

Comment (CarloLucibello) [inline base/special/gamma.jl]: I  agree, so better stay with Int64 conversion. In any case one could always resort to the BigInt version of `binomial` if needs to

Comment (tkelman) [inline base/special/gamma.jl]: If you know when it's appropriate to overflow I think it's better to do so rather than throw an `InexactError` in `round(Int64, n)`. This also needs quite a few more tests.

File: base/special/gamma.jl [hunk 2]
```diff
@@ -482,3 +482,7 @@ eta(x::Integer) = eta(Float64(x))
...
+                                ? convert(promote_type(T,S), binomial(round(Int64, n), round(Int64, k)))
+                                : gamma(1+n) / (gamma(1+n-k) * gamma(1+k)) )
```
Comment (stevengj) [inline base/special/gamma.jl]: This seems like it is susceptible to spurious overflow.  It might be more reliable to do `exp(lgamma(1+n) - lgamma(1+n-k) - lgamma(1+k))`?

I wouldn't bother with the integer check and `binomial` calls, which are even more susceptible to overflow.  If the user has integer values and wants specialized code, they should convert to integers themselves.

Comment (andreasnoack) [inline base/special/gamma.jl]: This is how we compute the [Beta function](https://github.com/JuliaLang/julia/blob/f308703afc24d4486d46efeec7b4088e825457c9/base/special/gamma.jl#L405-L411) so I think you should just call `beta` here.

Comment (stevengj) [inline base/special/gamma.jl]: @andreasnoack, the beta function is similar but does not seem identical.  If `x = 1 + n - k` and `w = 1 + k`, then `x + w` = 2 + n` and hence the function here does not seem to match `1/beta(x,w)`.

Comment (andreasnoack) [inline base/special/gamma.jl]: I think `binomial(x,y) = inv((x+1)*beta(x - y + 1, y + 1))` should do the job.

Comment (stevengj) [inline base/special/gamma.jl]: True.

Comment (CarloLucibello) [inline base/special/gamma.jl]: >I think binomial(x,y) = inv((x+1)*beta(x - y + 1, y + 1)) should do the job

This looks like a nice solution. 

>I wouldn't bother with the integer check and binomial calls, which are even more susceptible to overflow. If the user has integer values and wants specialized code, they should convert to integers themselves.

I would like to cover cases such as
```julia
julia> f(x,y) = inv((x+1)*beta(x - y + 1, y + 1))
f (generic function with 1 method)

julia> binomial(-2.,4.)
5.0

julia> f(-2.,4.)
NaN
```

I could just do the residues calculation without casting to integers....

Comment (stevengj) [inline base/special/gamma.jl]: @CarloLucibello, I think that is a bug in our `beta` function.  `beta(-n,n)` should return `-1/n`, not NaN.

Comment (stevengj) [inline base/special/gamma.jl]: See #14256.

Comment (stevengj) [inline base/special/gamma.jl]: In fact, the kinds of subtleties handled by the Cephes implementation I linked from 14256 make it even clearer that the `binomial` function should call `beta` for non-`Integer` arguments — we don't want to implement these subtleties twice.

**Comment by andreasnoack** [comment]
@CarloLucibello It would be great to have this PR completed.

**Comment by CarloLucibello** [comment]
@andreasnoack Right, I'll try to complete it during this week

**Comment by musm** [comment]
> In fact, the kinds of subtleties handled by the Cephes implementation I linked from 14256 make it even clearer that the binomial function should call beta for non-Integer arguments — we don't want to implement those subtleties twice.

Since the beta special function has been moved to SpecialFunctions, addressing the issue here requires the method to be added in that package.
